### PR TITLE
Document passing in debugger commands

### PR DESCRIPTION
**Pull request by dschwen**
This made parallel debugging a lot more convenient for me.

Refs #11170

### CODE REVIEW

**Review by permcody — APPROVED** [review]

File: framework/doc/content/documentation/application_development/debugging.md [hunk 1]
```diff
@@ -126,6 +126,16 @@ to be able to stop if MOOSE encounters an error.
...
 
+##### Passing debugger commands on the command line
+
+`lldb` accepts debugger commands through the `-o` command line option that are executed as soon as the execuatble is loaded up and ready. This can be used to set breakpoints and immediately resume the execution of the app.
+
+```bash
+mpirun -n 4 ./yourapp-dbg -i inputfile.i --start-in-debugger "sudo lldb -o 'break set -E C++' -o cont"
+```
+
+The above command will set breakpoints that halt on thrown exceptions. Replace `break set -E C++` with `b MPI_Abort` to break on MOOSE errors. The `-o cont` option will automatically run the app after teh breakpoints are set.
```
Comment (permcody) [inline framework/doc/content/documentation/application_development/debugging.md]: s/teh/the/

**Review by permcody — CHANGES_REQUESTED** [review]

File: framework/doc/content/documentation/application_development/debugging.md (hunk 1 above)
Comment (dschwen) [reply framework/doc/content/documentation/application_development/debugging.md]: ARGH!!!

**Review by permcody — APPROVED** [review]

**Comment by dschwen** [comment]
Btw. I put a `sudo` in the docs to appease @friedmud. I run this without it and don't enter a password anywhere. My computer must be sprinkled with magic fairy dust... or the new moose environment package I have installed really does _not_ require sudo anymore... 

¯\_(ツ)_/¯

**Comment by jwpeterson** [comment]
@dschwen is `which lldb` == `/usr/bin/lldb` for you?

This one does not require me to type my password for some reason...

**Comment by permcody** [comment]
> This one does not require me to type my password for some reason...

The some reason is because the lldb distributed from Apple prompts you on-screen for the password the very first time you try to run it and then never requires a password again. @milljm hasn't figured out how to make that work for the one he supplies in the MOOSE-package. It would be cool if that did work. I routinely find myself forgetting sudo and then the debugger just doesn't work (even though it runs).

**Comment by permcody** [comment]
It was a joke. We genuinely want to know why that works when it doesn't work for anyone else who's using the new package. That's the place we'd like to reach. 

So are you using our lldb or the system lldb? If you are using system lldb, then that explains it, if you are using our lldb, the next question is, which package number? If it's older, that would also explain it. It was more recently when we started building our own from source. The reason is because while the old lldb mostly works, we found a few specific cases where it fails to print out vector values (I can dig those up if you are interested).

So at the end of the day, I would expect that you would have the same problem as we do if you install the latest version of the package and use our lldb (not the system's). This is the problem we'd like to solve.

**Comment by dschwen** [comment]
I know @permcody, I'm just giving you a hard time. I also didn't see the comments after yours as I still had the page open in a stale tab.

Check this out:

```
> which lldb
/usr/bin/lldb
```

```
> find /opt/ | grep lldb
>
```

there is no `lldb` in `/opt/moose` or anywhere underneath `/opt`. What's up with that?

> and then never requires a password again.

Yeah, that's my observation. It didn't even ask me after the power outage last week (contrary to Derek's suggestion that mac lldb gets _"elevated privileges and it sticks for a while (I think 30 minutes by default)"_).

**Comment by milljm** [comment]
I don't know why I didn't see an email until today. Sorry guys...

we need sudo when using our version of lldb (/opt/moose).

...we still need sudo when using Apple's lldb. Sort of. Once you prove to Apple you're a super user (by means of a graphical pop-up prompt which in all intent and purposes _is_ sudo), it doesn't ask again... Ever (which means Its probably tied to the keychain somehow).

Ever since I have been imaging machines for you guys here at the lab, I have always (and trained the other techs here to do the same) added users to the '_developer' group:

```bash
  dscl . -append /Groups/admin GroupMemebership <your user id>  #<-- for sudo
  dscl . -append /Groups/_developer GroupMemebership <your user id> #<-- prevents further need of sudo
```

If you all check your account via `id` you will see you are apart of these groups (namely the _developers group) [You can read about this group, and what it does here](https://forums.developer.apple.com/thread/44758).

**Comment by permcody** [comment]
```
$ which lldb
/opt/moose/llvm-5.0.1/bin/lldb
```

**Comment by dschwen** [comment]
Could this be relevant, too?

https://gcc.gnu.org/onlinedocs/gcc-4.8.1/gnat_ugn_unw/Codesigning-the-Debugger.html

**Comment by milljm** [comment]
@dschwen Unfortunately I don't think we can 're-distribute' the type of certificate required (Mac Development) to tell all machines using our binaries that they do not require sudo to do sudo things on other peoples machines :) That would be awesome.. and scary right?

We do own, and can distribute, a Developer ID Installer certificate. Which tells the OS "Yes. This app is trusted and can be installed... so long as YOU have sudo rights to do so".

Reference: https://help.apple.com/xcode/mac/current/#/dev80c6204ec

**Comment by milljm** [comment]
I have tried to look into this in the past, and still come to the same conclusion today. That it aint easy.

https://stackoverflow.com/questions/28542582/xcode-custom-lldb-or-adding-args-to-existing-one
http://lldb.llvm.org/build.html#BuildingLldbOnMacOSX

The latter 'Official' instructions from LLVM.org are explaining how to _replace_ a users system (Xcode) LLDB. That is something that will get us murdered by other Mac developers.

While I agree it would be nice to figure out how to run lldb without sudo, I do not know how. Not with out breaking some serious security holes.

**Comment by friedmud** [comment]
I’m not personally bothered by the sudo - since that’s effectively what’s
happening with Apple’s binary anyway...

I just want our directions to be consistent with our package.

Derek

**Comment by permcody** [comment]
The only problem with forcing sudo is that it's not clear why it's not working when you forget. I'd rather have a message that says "run with sudo" instead of just not working.

**Comment by permcody** [comment]
Jason may have solved this - Andrew is reporting that it's working after the newest, newest package and a reboot.

**Comment by milljm** [comment]
There is more to it than this... no one should be installing the latest package to get access to non-sudo lldb. I still need to work out the installation portion of it all. I am also not sure what happens when we do certain things twice. Or when there is a new version of LLDB being built do we need to re-install the code-signing cert thats in play. There is also the issue with the cert only lasting one year. Can this be extended or, do we even care because we pump out multiple packages a year, etc etc...

But, I did get it working yes.

**Pull request merged**